### PR TITLE
Hardware JPEG Compression via VAAPI

### PR DESCRIPTION
**Pull request by TrentHouliston**
This adds code that will perform the compression of JPEG images using the intel integrated GPU. This will take load off the CPU and allow for faster compression rates for saving or sending to NUsight.
One downside is that this may compete with the VisualMesh for GPU resources and this will need to be factored in.

Also this code no longer does debayering for jpeg compression. It instead performs a permutation of the Bayer to make it easier to compress.
The resulting image is a black an white image which changes the order like so
```
RGRGRGRG  |  RRRRGGGG
GBGBGBGB  |  RRRRGGGG
RGRGRGRG  |  GGGGBBBB
GBGBGBGB  |  GGGGBBBB
```

NUsight2 will need to be modified with new shaders that can handle this format.

### CODE REVIEW

File: module/output/ImageCompressor/src/compressor/vaapi/operation/upload_to_surface.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        case utility::vision::fourcc("RGGB"):
+        case utility::vision::fourcc("GRBG"):
+        case utility::vision::fourcc("GBRG"):
+        case utility::vision::fourcc("PBG8"):
+        case utility::vision::fourcc("PRG8"):
+        case utility::vision::fourcc("PGR8"):
+        case utility::vision::fourcc("PGB8"):
+        case utility::vision::fourcc("GRAY"):
+        case utility::vision::fourcc("GREY"): {
+            // Simple version pitch matches width we can just copy
```
Comment (Bidski) [inline module/output/ImageCompressor/src/compressor/vaapi/operation/upload_to_surface.cpp]: Grammar!

File: nuclear/cmake/Modules/NUClearCompilerSettings.cmake [hunk 2]
```diff
@@ -20,6 +20,5 @@ set(CMAKE_MACOSX_RPATH ON)
 set(CMAKE_INSTALL_RPATH ${CMAKE_INSTALL_RPATH} lib/ ../lib/ bin/lib)
 
 if(NOT MSVC)
-  # Compilation must be done with c++14 for NUClear to work
-  add_compile_options(-std=c++14 -fPIC)
+  add_compile_options(-std=c++17 -fPIC)
```
Comment (Bidski) [inline nuclear/cmake/Modules/NUClearCompilerSettings.cmake]: [CMAKE_CXX_STANDARD](https://cmake.org/cmake/help/v3.1/variable/CMAKE_CXX_STANDARD.html#variable:CMAKE_CXX_STANDARD)
[CMAKE_POSITION_INDEPENDENT_CODE](https://cmake.org/cmake/help/v3.0/variable/CMAKE_POSITION_INDEPENDENT_CODE.html#variable:CMAKE_POSITION_INDEPENDENT_CODE)

File: module/output/ImageCompressor/src/compressor/vaapi/cl/opencl_context_for_display.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                cl_int err = devices_from_display(platform_id,
+                                                  CL_VA_API_DISPLAY_INTEL,
+                                                  dpy,
+                                                  CL_PREFERRED_DEVICES_FOR_VA_API_INTEL,
+                                                  0,
+                                                  nullptr,
+                                                  &n_devices);
+
+                // If there was a device on this platform get it
+                if (err == CL_SUCCESS && len > 0) {
```
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply module/output/ImageCompressor/src/compressor/vaapi/cl/opencl_context_for_display.cpp]: Totally is!

File: module/output/ImageCompressor/src/compressor/vaapi/cl/opencl_context_for_display.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                        clGetExtensionFunctionAddressForPlatform(platform_id,
+                                                                 "clEnqueueAcquireVA_APIMediaSurfacesINTEL"));
+                    auto release_surface = reinterpret_cast<clEnqueueReleaseVA_APIMediaSurfacesINTEL_fn>(
+                        clGetExtensionFunctionAddressForPlatform(platform_id,
+                                                                 "clEnqueueReleaseVA_APIMediaSurfacesINTEL"));
+
+                    if (mem_from_surface == nullptr || acquire_surface == nullptr || release_surface == nullptr) {
+                        throw std::runtime_error(
+                            "Unable to get the function pointer for converting surfaces to mem "
+                            "objects");
```
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply module/output/ImageCompressor/src/compressor/vaapi/cl/opencl_context_for_display.cpp]: For some reason, I terminated, and then unterminated the string

File: module/output/ImageCompressor/src/ImageCompressor.cpp [hunk 5]
```diff
@@ -1,101 +1,184 @@
...
+        on<Every<1, std::chrono::seconds>>().then("Stats", [this] {
+            log<NUClear::INFO>(fmt::format("Receiving {}/s, Dropping {}/s ({}%)",
+                                           compressed + dropped,
+                                           dropped,
+                                           100 * double(compressed) / double(compressed + dropped)));
+            compressed = 0;
+            dropped    = 0;
+        });
+
+        on<Every<2000, Per<std::chrono::seconds>>>().then([this] {
```
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply module/output/ImageCompressor/src/ImageCompressor.cpp]: It was there to test the performance of the compressor. It won't be here for merge.

File: module/output/ImageCompressor/src/compressor/vaapi/operation/upload_to_surface.cpp (hunk 1 above)
Comment (TrentHouliston) [reply module/output/ImageCompressor/src/compressor/vaapi/operation/upload_to_surface.cpp]: I English good have!

File: nuclear/cmake/Modules/NUClearCompilerSettings.cmake (hunk 2 above)
Comment (TrentHouliston) [reply nuclear/cmake/Modules/NUClearCompilerSettings.cmake]: Really our whole method of handling cxx flags and features needs to change, and this is the minimal change required to fix this. If you want to fix this properly it should be in another PR

File: module/output/ImageCompressor/src/compressor/vaapi/cl/make_mosaic_kernel.cpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+#include "opencl_error_category.h"
+
+namespace module::output::compressor::vaapi::cl {
+std::pair<cl::program, cl::kernel> make_mosaic_kernel(CompressionContext::OpenCLContext context,
+                                                      const uint32_t& width,
+                                                      const uint32_t& height,
+                                                      const uint32_t& format) {
+    // Make the OpenCL source
+    int mosaic_size    = mosaic::mosaic_size(format);
+    std::string source = fmt::format(R"(
```
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply module/output/ImageCompressor/src/compressor/vaapi/cl/make_mosaic_kernel.cpp]: It's fast enough, this is something we can look at if the OpenCL render speed is ever the weak link

File: module/output/ImageCompressor/src/compressor/turbojpeg/Compressor.cpp [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                    case utility::vision::fourcc("GRBG"):
+                    case utility::vision::fourcc("GBRG"):
+                    case utility::vision::fourcc("PY  "):
+                    case utility::vision::fourcc("PBG8"):
+                    case utility::vision::fourcc("PRG8"):
+                    case utility::vision::fourcc("PGR8"):
+                    case utility::vision::fourcc("PGB8"):
+                    case utility::vision::fourcc("Y8  "):
+                    case utility::vision::fourcc("GRAY"):
+                    case utility::vision::fourcc("GREY"):
```
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply module/output/ImageCompressor/src/compressor/turbojpeg/Compressor.cpp]: Handled

File: module/output/ImageCompressor/CMakeLists.txt [hunk 8]
```diff
@@ -3,6 +3,8 @@ find_package(va REQUIRED)
 find_package(va-drm REQUIRED)
 find_package(OpenCL REQUIRED)
 
+add_compile_definitions(CL_TARGET_OPENCL_VERSION=120)
```
[the remark this answers is not in the document]
Comment (Bidski) [reply module/output/ImageCompressor/CMakeLists.txt]: > Adds preprocessor definitions to the compiler command line for targets in the current directory and below

[add_compile_definitions](https://cmake.org/cmake/help/latest/command/add_compile_definitions.html)

File: module/output/ImageCompressor/src/ImageCompressor.cpp [hunk 9]
```diff
@@ -1,101 +1,184 @@
...
-                case fourcc("GBRG"): compress(debayer(image, method), TJPF_RGB); break;
-
-                // Regular convertible formats
-                case fourcc("RGB3"):
-                case fourcc("RGB8"): compress(image, TJPF_RGB); break;
-                case fourcc("BGR8"): compress(image, TJPF_BGR); break;
-                case fourcc("GRAY"):
-                case fourcc("GREY"): compress(image, TJPF_GRAY); break;
-            }
+        on<Every<1, std::chrono::seconds>>().then("Stats", [this] {
```
[the remark this answers is not in the document]
Comment (Bidski) [reply module/output/ImageCompressor/src/ImageCompressor.cpp]: This is what the config file is for

File: cmake/Modules/Findva-drm.cmake [hunk 10]
```diff
@@ -0,0 +1,2 @@
+include(ToolchainLibraryFinder)
```
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply cmake/Modules/Findva-drm.cmake]: It is a component of libva, sorta and I'm not sure how to add components functionality

File: nuclear/cmake/Modules/NUClearCompilerSettings.cmake [hunk 11]
```diff
@@ -19,6 +19,9 @@ set(CMAKE_MACOSX_RPATH ON)
...
 
-set(CMAKE_CXX_STANDARD 17)
-set(CMAKE_CXX_STANDARD_REQUIRED ON)
-set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+if(NOT MSVC)
+  # Compilation must be done with c++17 for NUClear to work
+  set(CMAKE_CXX_STANDARD 17)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+  set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+endif()
```
Comment (Bidski) [inline nuclear/cmake/Modules/NUClearCompilerSettings.cmake]: This should come in from master

Comment (TrentHouliston) [reply nuclear/cmake/Modules/NUClearCompilerSettings.cmake]: It did, you're commenting on an outdated diff

**Comment by TrentHouliston** [comment]
This goes along with https://github.com/NUbots/NUsight2/pull/298 which allows NUsight2 to decode and display the images properly

**Pull request merged**